### PR TITLE
Rename `TypeSignature::NullAry` --> `TypeSignature::Nullary` and improve comments

### PR DESCRIPTION
**Pull request by alamb**
## Which issue does this PR close?

- Part of https://github.com/apache/datafusion/issues/13763

## Rationale for this change
- @kylebarron points out in https://github.com/apache/datafusion/pull/13354#discussion_r1887730438 that:

> Is the A here intended to be upper-case? I'd have expected Nullary

While I was making that change I noticed that the `TypeSignature` documentation could be improved as well so I did it at the same time

## What changes are included in this PR?
1. Rename `TypeSignature::NullAry` --> `TypeSignature::Nullary` 
2.  improve comments

## Are these changes tested?
By CI

## Are there any user-facing changes?
The name of an enum has changed, but this has not yet been released, so it isn't a user visible change in my mind

### CODE REVIEW

File: datafusion/expr-common/src/signature.rs [hunk 1]
```diff
@@ -74,6 +74,8 @@ pub enum Volatility {
 /// adds a cast such as `cos(CAST int_column AS DOUBLE)` during planning.
 ///
 /// # Data Types
+///
+/// ## Timestamps
```
Comment (findepi) [inline datafusion/expr-common/src/signature.rs:78]: should there be one empty line after a header?

File: datafusion/expr-common/src/signature.rs [hunk 2]
```diff
@@ -102,51 +106,75 @@ pub enum TypeSignature {
...
+    /// One or more arguments of an arbitrary but equal type out of a list of valid types.
     ///
     /// # Examples
     /// 1. A function of one argument of f64 is `Uniform(1, vec![DataType::Float64])`
     /// 2. A function of one argument of f64 or f32 is `Uniform(1, vec![DataType::Float32, DataType::Float64])`
     Uniform(usize, Vec<DataType>),
-    /// Exact number of arguments of an exact type
+    /// One or more arguments with exactly the specified types in order.
+    ///
+    /// For functions that take no arguments (e.g. `random()` see [`TypeSignature::Nullary`]).
```
Comment (findepi) [inline datafusion/expr-common/src/signature.rs:117]:
```suggestion
    /// For functions that take no arguments (e.g. `random()`) use [`TypeSignature::Nullary`].
```

File: datafusion/expr-common/src/signature.rs [hunk 3]
```diff
@@ -102,51 +106,75 @@ pub enum TypeSignature {
...
     Exact(Vec<DataType>),
-    /// The number of arguments that can be coerced to in order
+    /// One or more arguments belonging to the [`TypeSignatureClass`], in order.
+    ///
     /// For example, `Coercible(vec![logical_float64()])` accepts
     /// arguments like `vec![DataType::Int32]` or `vec![DataType::Float32]`
-    /// since i32 and f32 can be casted to f64
+    /// since i32 and f32 can be cast to f64
+    ///
+    /// For functions that take no arguments (e.g. `random()` see [`TypeSignature::Nullary`]).
```
Comment (findepi) [inline datafusion/expr-common/src/signature.rs:125]: here same
```suggestion
    /// For functions that take no arguments (e.g. `random()`) see [`TypeSignature::Nullary`].
```

File: datafusion/expr-common/src/signature.rs [hunk 4]
```diff
@@ -102,51 +106,75 @@ pub enum TypeSignature {
...
     Coercible(Vec<TypeSignatureClass>),
-    /// The arguments will be coerced to a single type based on the comparison rules.
-    /// For example, i32 and i64 has coerced type Int64.
+    /// One or more arguments that can be "compared"
```
Comment (findepi) [inline datafusion/expr-common/src/signature.rs:127]: what does this mean ?

File: datafusion/expr-common/src/signature.rs [hunk 5]
```diff
@@ -102,51 +106,75 @@ pub enum TypeSignature {
...
+    ///
+    /// Each argument will be coerced to a single type based on comparison rules.
+    /// For example a function called with `i32` and `i64` has coerced type `Int64` so
+    /// each argument will be coerced to `Int64` before the function is invoked.
     ///
     /// Note:
-    /// - If compares with numeric and string, numeric is preferred for numeric string cases. For example, nullif('2', 1) has coerced types Int64.
+    /// - If compares with numeric and string, numeric is preferred for numeric string cases. For example, `nullif('2', 1)` has coerced types `Int64`.
```
Comment (findepi) [inline datafusion/expr-common/src/signature.rs:134]: is the reader expected to know/assume that nullif uses signature "Comparable"?

File: datafusion/expr-common/src/signature.rs [hunk 6]
```diff
@@ -102,51 +106,75 @@ pub enum TypeSignature {
...
+    /// - See [`comparison_coercion`] for more details.
+    /// - For functions that take no arguments (e.g. `random()` see [`TypeSignature::Nullary`]).
     ///
-    /// See `comparison_coercion_numeric` for more details.
+    /// [`comparison_coercion`]: crate::type_coercion::binary::comparison_coercion
     Comparable(usize),
-    /// Fixed number of arguments of arbitrary types, number should be larger than 0
+    /// One or more arguments of arbitrary types.
+    ///
+    /// For functions that take no arguments (e.g. `random()` see [`TypeSignature::Nullary`]).
```
Comment (findepi) [inline datafusion/expr-common/src/signature.rs:143]:
```suggestion
    /// For functions that take no arguments (e.g. `random()`) use [`TypeSignature::Nullary`].
```

File: datafusion/expr-common/src/signature.rs [hunk 7]
```diff
@@ -102,51 +106,75 @@ pub enum TypeSignature {
...
+    /// Matches exactly one of a list of [`TypeSignature`]s.
+    ///
+    /// Coercion is attempted to match the signatures in order, and stops after
+    /// the first success, if any.
     ///
     /// # Examples
-    /// Function `make_array` takes 0 or more arguments with arbitrary types, its `TypeSignature`
+    ///
+    /// Since `make_array` takes 0 or more arguments with arbitrary types, its `TypeSignature`
     /// is `OneOf(vec![Any(0), VariadicAny])`.
```
Comment (findepi) [inline datafusion/expr-common/src/signature.rs:153]: This IMO is wrong design of those signatures, but whatever (for now)

File: datafusion/expr-common/src/signature.rs [hunk 8]
```diff
@@ -102,51 +106,75 @@ pub enum TypeSignature {
...
-    /// Specifies Signatures for array functions
+    /// A function that has an [`ArrayFunctionSignature`]
     ArraySignature(ArrayFunctionSignature),
-    /// Fixed number of arguments of numeric types.
+    /// One or more arguments of numeric types.
+    ///
     /// See [`NativeType::is_numeric`] to know which type is considered numeric
     ///
-    /// [`NativeType::is_numeric`]: datafusion_common
+    /// For functions that take no arguments (e.g. `random()` see [`TypeSignature::Nullary`]).
```
Comment (findepi) [inline datafusion/expr-common/src/signature.rs:161]:
```suggestion
    /// For functions that take no arguments (e.g. `random()`) use [`TypeSignature::Nullary`].
```

File: datafusion/expr-common/src/signature.rs [hunk 9]
```diff
@@ -102,51 +106,75 @@ pub enum TypeSignature {
...
+    /// One or arguments of all the same string types.
+    ///
     /// The precedence of type from high to low is Utf8View, LargeUtf8 and Utf8.
     /// Null is considered as `Utf8` by default
     /// Dictionary with string value type is also handled.
+    ///
+    /// For example, if a function is called with (utf8, large_utf8), all
+    /// arguments will be coerced to  `LargeUtf8`
+    ///
+    /// For functions that take no arguments (e.g. `random()` see [`TypeSignature::Nullary`]).
```
Comment (findepi) [inline datafusion/expr-common/src/signature.rs:174]: here too

**Review by findepi — APPROVED** [review]

File: datafusion/expr-common/src/signature.rs (hunk 5 above)
Comment (jayzhan211) [reply datafusion/expr-common/src/signature.rs:134]: actual example direct reader to understand this easily

**Comment by jayzhan211** [comment]
Thanks @alamb

File: datafusion/expr-common/src/signature.rs (hunk 1 above)
Comment (alamb) [reply datafusion/expr-common/src/signature.rs:78]: Good call -- fixed

File: datafusion/expr-common/src/signature.rs [hunk 10]
```diff
@@ -93,60 +96,89 @@ pub enum Volatility {
 pub enum TypeSignature {
     /// One or more arguments of a common type out of a list of valid types.
     ///
+    /// For functions that take no arguments (e.g. `random()` see [`TypeSignature::Nullary`]).
+    ///
     /// # Examples
-    /// A function such as `concat` is `Variadic(vec![DataType::Utf8, DataType::LargeUtf8])`
+    ///
+    /// A function such as `concat` is `Variadic(vec![DataType::Utf8,
+    /// DataType::LargeUtf8])`
```
Comment (findepi) [inline datafusion/expr-common/src/signature.rs:104]: nit:
Maybe it would be more readable to go without `DataType:` for type names?

**Review by findepi — APPROVED** [review]

File: datafusion/expr-common/src/signature.rs (hunk 4 above)
Comment (findepi) [reply datafusion/expr-common/src/signature.rs:131]: this needs a follow-up

Comment (alamb) [reply datafusion/expr-common/src/signature.rs:131]: Agreed -- this can be made better. I will attempt a follow up

**Comment by alamb** [comment]
Thank you for the reviews @jayzhan211 and @findepi  -- I think this PR is an improvement over what is on main (though there is still more improvement to be had). Thus merging it in and will work on additional clarifications as a follow up

File: datafusion/expr-common/src/signature.rs (hunk 4 above)
Comment (alamb) [reply datafusion/expr-common/src/signature.rs:131]: - Next round PR https://github.com/apache/datafusion/pull/13840

File: datafusion/expr-common/src/signature.rs (hunk 10 above)
Comment (alamb) [reply datafusion/expr-common/src/signature.rs:104]: in https://github.com/apache/datafusion/pull/13840

**Pull request merged**